tensorflow/core/kernels/data/concatenate_dataset_op_test.cc

```diff
@@ -0,0 +1,378 @@
+                             TensorShape{2, 3}, {1, 2, 3, 4, 5, 6}),
+                         DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {7, 8, 9, 10})},
+                        {DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 2}, {11, 12, 13, 14}),
+                         DatasetOpsTestBase::CreateTensor<int64>(
+                             TensorShape{2, 1}, {15, 16})}},
+     /*expected_outputs*/
+     {DatasetOpsTestBase::CreateTensor<int64>(TensorShape{3}, {1, 2, 3}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {7, 8}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{3}, {4, 5, 6}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {9, 10}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {11, 12}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{1}, {15}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{2}, {13, 14}),
+      DatasetOpsTestBase::CreateTensor<int64>(TensorShape{1}, {16})},
+     /*expected_output_dtypes*/ {DT_INT64, DT_INT64},
+     /*expected_output_shapes*/
+     {PartialTensorShape({-1}), PartialTensorShape({-1})},
+     /*expected_cardinality*/ 4,
+     /*breakpoints*/ {0, 2, 5}}};
+
+struct ConcatenateDatasetOpTestHelper : ConcatenateDatasetOpTest {
+  ~ConcatenateDatasetOpTestHelper() {
+    if (dataset) dataset->Unref();
+  }
+
+  Status CreateDatasetFromTestCase(const TestParam &test_case) {
+    std::vector<Tensor> tensor_slice_dataset_tensors;
+    TF_RETURN_IF_ERROR(CreateTensorSliceDatasetTensors(
+        test_case.input_tensors, &tensor_slice_dataset_tensors));
+    gtl::InlinedVector<TensorValue, 4> inputs;
+    for (auto &tensor : tensor_slice_dataset_tensors) {
+      inputs.emplace_back(&tensor);
+    }
+    TF_RETURN_IF_ERROR(CreateConcatenateDatasetKernel(
+        test_case.expected_output_dtypes, test_case.expected_output_shapes,
+        &dataset_kernel));
+    TF_RETURN_IF_ERROR(CreateConcatenateDatasetContext(
+        dataset_kernel.get(), &inputs, &dataset_kernel_ctx));
+    TF_RETURN_IF_ERROR(CreateDataset(dataset_kernel.get(),
+                                     dataset_kernel_ctx.get(), &dataset));
+    return Status::OK();
+  }
+
+  Status CreateIteratorFromTestCase(const TestParam &test_case) {
+    TF_RETURN_IF_ERROR(CreateDatasetFromTestCase(test_case));
+    TF_RETURN_IF_ERROR(
+        CreateIteratorContext(dataset_kernel_ctx.get(), &iterator_ctx));
+    TF_RETURN_IF_ERROR(
+        dataset->MakeIterator(iterator_ctx.get(), "Iterator", &iterator));
+    return Status::OK();
+  }
+
+  std::unique_ptr<OpKernel> dataset_kernel;
+  std::unique_ptr<OpKernelContext> dataset_kernel_ctx;
+  DatasetBase *dataset = nullptr;  // owned by this class.
+  std::unique_ptr<IteratorContext> iterator_ctx;
+  std::unique_ptr<IteratorBase> iterator;
+};
+
+struct DatasetGetNextTest : public ConcatenateDatasetOpTestHelper,
```

Comment:
"Use a `struct` only for passive objects that carry data; everything else is a `class`." https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes